libpgcommon/lwgeom_cache.c

```diff
@@ -303,25 +292,18 @@ ToastCacheGetGeometry(FunctionCallInfo fcinfo, uint32_t argnum)
 	/* We've seen this object before? */
 	if (arg->valueid == valueid && arg->toastrelid == toastrelid)
 	{
-		if (arg->geom)
```

Comment:
The point of only loading on a cache hit (counterintuitively) is so that we only load when we've see the object *twice*. That way we "know" (loosely) that we're in a repeating pattern. Loading on every cache miss can mean taking a copy every time, in that case where we aren't seeing a repeating pattern. I figure the trade off in the coherent case (one extra load at the start) was offset by the incoherent case (every load is a wasted load).

Reply:
The truth is, I did this because with the old behaviour when ToastCacheGetGeometry sees the toasted geometry for the first time it would return a temporal shared shared_gserialized (same as the non toasted value). When this reaches the GetGeomCache, it wants to cache it, so it adds a reference (which in turn will do a copy in the cache context, since it needs it to live). 
The second time the geometry appears, the toastcache will create a "perpetual" shared_gserialized, but when this reaches the getgeocache the geometry is already in cache, so you have 2 "perpetual" shared_gserialized containing different copies of the geometry but they pass the equality check so they won't "merge" (and perform the memcpy everytime).
There are several ways to handle this, like updating the cache if the gserialized objects are equal in value but not in pointer, but removing the possibility altogether seemed like a simpler solution.

> Loading on every cache miss can mean taking a copy every time, in that case where we aren't seeing a repeating pattern.

Keep in mind that when we don't "persist" on the cache context here, we need to copy it in the GetGeomCach. Also, by using this shared cache we are only keeping a single copy of gserialized object, and not 2 like before.

Talking from memory, I think that PG_DETOAST_DATUM and PG_DETOAST_DATUM_COPY will act the same for toasted datums (you never get a non copied value since it doesn't exist on disk), so there shouldn't be any change performance wise if that worries you.